resolwe_bio/processes/import_data/geneset.py

```diff
@@ -0,0 +1,316 @@
+"""Upload geneset."""
+import gzip
+import json
+from pathlib import Path
+
+from resolwe.process import (
+    FileField,
+    JsonField,
+    ListField,
+    Persistence,
+    Process,
+    SchedulingClass,
+    StringField,
+)
+from resolwe.process.models import DescriptorSchema
+
+
+def parse_geneset_file(geneset_file, warning):
+    """Parse geneset file."""
+    with open(geneset_file, "rU") as handle:
+        # skip empty lines
+        genes = [str(line.strip()) for line in handle if line.strip()]
+        geneset = sorted(set(genes))
+        if len(genes) != len(geneset):
+            warning("Removed duplicated genes.")
+
+    return geneset
+
+
+def save_geneset_to_json(geneset, output_json):
+    """Save geneset to json file."""
+    with open(output_json, "w") as handle:
+        json.dump(
+            {"genes": geneset},
+            handle,
+            separators=(",", ":"),
+            allow_nan=False,
+        )
+
+
+def save_geneset_to_file(geneset, output_file):
+    """Save geneset to file."""
+    with gzip.open(output_file, "w") as handle:
+        handle.write("\n".join(geneset).encode("utf-8"))
+
+
+class UploadGeneset(Process):
+    """
+    Upload a set of genes.
+
+    Provide one gene ID per line in a .tab, .tab.gz, or .txt file format.
+    """
+
+    slug = "upload-geneset"
+    name = "Gene set"
+    process_type = "data:geneset"
+    version = "1.3.0"
+    category = "Import"
+    scheduling_class = SchedulingClass.INTERACTIVE
+    persistence = Persistence.RAW
+    requirements = {
+        "expression-engine": "jinja",
+        "executor": {
+            "docker": {
+                "image": "public.ecr.aws/s4q6j6e8/resolwebio/base:ubuntu-20.04-03042021"
+            }
+        },
+        "resources": {
+            "cores": 1,
+            "memory": 1024,
+        },
+    }
+    data_name = '{{ src.file|default("?") }}'
+
+    class Input:
+        """Input fields."""
+
+        src = FileField(
+            label="Gene set",
+            description="List of genes (.tab/.txt, one Gene ID per line. "
+            "Supported extensions: .tab, .tab.gz (preferred), tab.*",
```

Comment:
Maybe add `txt` or/and `txt.*` to supported extensions, since it was mentioned before that it also takes `.txt` file format.